File: java/src/main/java/libldt3/model/regel/kontext/KontextregelHelper.java
/*
 * Copyright 2016-2022  Christoph Brill <opensource@christophbrill.de>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package libldt3.model.regel.kontext;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import libldt3.annotations.Feld;
import libldt3.annotations.Objekt;
import libldt3.model.Kontext;
import libldt3.model.objekte.Fliesstext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

class KontextregelHelper {

    private static final Logger LOG = LoggerFactory.getLogger(KontextregelHelper.class);

    /**
     * Check if a given field has any string content (either simply text or multiline Fliesstext)
     */
    public static boolean containsAnyString(Field field, Object owner) throws IllegalAccessException {
        Object value = getFieldValue(field, owner);
        if (value == null) return false;
        return containsAnyString(value);
    }

    public static Object getFieldValue(Field field, Object owner) throws IllegalAccessException {
        if (field == null) {
            LOG.warn("No field given, cannot check for content");
            return null;
        }
        field.setAccessible(true);
        return field.get(owner);
    }

    /**
     * Check if a given field has any string content (either simply text or multiline Fliesstext)
     */
    public static boolean containsAnyString(Object value) throws IllegalAccessException {
        if (value instanceof String) {
            String o = (String) value;
            return !o.isEmpty();
        }
        if (value instanceof Fliesstext) {
            Fliesstext fliesstext = ((Fliesstext) value);
            if (fliesstext.text != null) {
                for (String s : fliesstext.text) {
                    if (s != null && !s.isEmpty()) {
                        return true;
                    }
                }
            }
            if (fliesstext.base64text != null) {
                for (String s : fliesstext.base64text) {
                    if (s != null && !s.isEmpty()) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    /**
     * Find a field matching their {@link Feld#value()} with the given fieldtype. Not recursive, only a single field.
     */
    public static Field findField(Object owner, String fieldtype) {
        for (Field f : owner.getClass().getDeclaredFields()) {
            Feld annotation = f.getAnnotation(Feld.class);
            if (annotation != null && annotation.value().equals(fieldtype)) {
                f.setAccessible(true);
                return f;
            }
        }
        return null;
    }

    /**
     * Find fields matching their {@link Feld#value()} with the given fieldtypes. Not recursive, but for multiple field.
     */
    public static Map<String, Field> findFields(Kontext owner, Set<String> fieldtypes) {
        Map<String, Field> result = new HashMap<>(fieldtypes.size());
        for (Field f : owner.getClass().getDeclaredFields()) {
            Feld annotation = f.getAnnotation(Feld.class);
            if (annotation != null && fieldtypes.contains(annotation.value())) {
                f.setAccessible(true);
                result.put(annotation.value(), f);
            }
        }
        return result;
    }

    /**
     * Find fields matching their {@link Feld#value()} with the given fieldtypes. Recursive and for multiple field.
     */
    public static Map<Kontext, Map<String, Field>> findFieldsRecursive(Kontext owner, Set<String> fieldtypes) throws IllegalArgumentException, IllegalAccessException {
        Map<Kontext, Map<String, Field>> result = new HashMap<>();
        Map<String, Field> fields = new HashMap<>();
        result.put(owner, fields);
        for (Field f : owner.getClass().getDeclaredFields()) {
            Feld annotation = f.getAnnotation(Feld.class);
            if (annotation != null && fieldtypes.contains(annotation.value())) {
                f.setAccessible(true);
                fields.put(annotation.value(), f);
            }
            f.setAccessible(true);
            Kontext obj = (Kontext) f.get(owner);
            if (obj != null) {
                if (obj.getClass().getAnnotation(Objekt.class) != null) {
                    result.putAll(findFieldsRecursive(obj, fieldtypes));
                } else if (obj instanceof Iterable<?>) {
                    for (Kontext o : (Iterable<Kontext>) obj) {
                        result.putAll(findFieldsRecursive(o, fieldtypes));
                    }
                }
            }
        }
        return result;
    }

}

File: cs/libldt3/model/regel/kontext/KontextregelHelper.cs
/*
 * Copyright 2016-2024  Christoph Brill <opensource@christophbrill.de>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
using System.Diagnostics;
using System.Reflection;
using libldt3.attributes;
using libldt3.model;
using libldt3.model.objekte;

namespace libldt3
{
    namespace model
    {
        namespace regel
        {
            namespace kontext
            {
                class KontextregelHelper
                {

                    public static bool ContainsAnyString(string value)
                    {
                        return !string.IsNullOrEmpty(value);
                    }
                    /// Check if a given field has any string content (either simply text or multiline Fliesstext)
                    public static bool ContainsAnyString(Fliesstext value)
                    {
                        Fliesstext fliesstext = (Fliesstext)value;
                        if (fliesstext.Text != null)
                        {
                            foreach (string s in fliesstext.Text)
                            {
                                if (s != null && !string.IsNullOrEmpty(s))
                                {
                                    return true;
                                }

                            }
                        }

                        if (fliesstext.Base64kodierteAnlage != null)
                        {
                            foreach (string s in fliesstext.Base64kodierteAnlage)
                            {
                                if (s != null && !string.IsNullOrEmpty(s))
                                {
                                    return true;
                                }

                            }
                        }

                        return false;
                    }
                    /// Check if a given field has any string content (either simply text or multiline Fliesstext)
                    public static bool ContainsAnyValue(FieldInfo field, object owner)
                    {
                        object value = KontextregelHelper.GetFieldValue(field, owner);
                        if (value == null)
                        {
                            return false;
                        }

                        if (value is string)
                        {
                            return KontextregelHelper.ContainsAnyString((string)value);
                        }

                        if (value is Fliesstext)
                        {
                            return KontextregelHelper.ContainsAnyString((Fliesstext)value);
                        }

                        return true;
                    }
                    /// Not recursive, only a single field.
                    public static FieldInfo FindField(object owner, string fieldtype)
                    {
                        foreach (FieldInfo f in owner.GetType().GetFields())
                        {
                            Feld? annotation = f.GetCustomAttribute<Feld>();
                            if (annotation != null && annotation.Value.Equals(fieldtype))
                            {
                                ;
                                return f;
                            }

                        }
                        return null;
                    }
                    /// Not recursive, but for multiple field.
                    public static IDictionary<string, FieldInfo> FindFields(Kontext owner, ISet<string> fieldtypes)
                    {
                        IDictionary<string, FieldInfo> result = new Dictionary<string, FieldInfo>(fieldtypes.Count);
                        foreach (FieldInfo f in owner.GetClass().GetFields())
                        {
                            Feld? annotation = f.GetCustomAttribute<Feld>();
                            if (annotation != null && fieldtypes.Contains(annotation.Value))
                            {
                                ;
                                result[annotation.Value] = f;
                            }

                        }
                        return result;
                    }
                    /// Recursive and for multiple field.
                    public static IDictionary<Kontext, IDictionary<string, FieldInfo>> FindFieldsRecursive(Kontext owner, ISet<string> fieldtypes)
                    {
                        IDictionary<Kontext, IDictionary<string, FieldInfo>> result = new Dictionary<Kontext, IDictionary<string, FieldInfo>>();
                        IDictionary<string, FieldInfo> fields = new Dictionary<string, FieldInfo>();
                        result[owner] = fields;
                        foreach (FieldInfo f in owner.GetClass().GetFields())
                        {
                            Feld? annotation = f.GetCustomAttribute<Feld>();
                            if (annotation != null && fieldtypes.Contains(annotation.Value))
                            {
                                ;
                                fields[annotation.Value] = f;
                            }

                    ;
                            object obj = f.GetValue(owner);
                            if (obj != null)
                            {
                                if (obj.GetType().GetCustomAttribute<Objekt>() != null)
                                {
                                    foreach (var x in KontextregelHelper.FindFieldsRecursive((Kontext)obj, fieldtypes)) { result[x.Key] = x.Value; };
                                }
                                else
                                {
                                    if (obj is IEnumerable<object>)
                                    {
                                        foreach (object o in (IEnumerable<Kontext>)obj)
                                        {
                                            if (o.GetType().GetCustomAttribute<Objekt>() != null)
                                            {
                                                foreach (var x in KontextregelHelper.FindFieldsRecursive((Kontext)o, fieldtypes)) { result[x.Key] = x.Value; };
                                            }

                                        }
                                    }

                                }

                            }

                        }
                        return result;
                    }
                    public static object GetFieldValue(FieldInfo field, object owner)
                    {
                        if (field == null)
                        {
                            Trace.TraceWarning("No field given, cannot check for content");
                            return null;
                        }

                    ;
                        return field.GetValue(owner);
                    }
                }
            }
        }
    }
}
